Refactor Deployment manifest to match the Statefulset manifest

mydoomfr opened the pull request:
We are currently using the Chart as Statefulset, but since we have no persistence, we're moving it to Deployment. 

Unfortunately, the Deployment is not working when Prometheus is enabled. The pod fails (CrashLoopBackOff) because the JMX exporter is missing.

```
[truncated]
sonarqube-sonarqube-5454cd7496-hfp9c sonarqube 2022.11.29 17:45:57 INFO  app[][o.s.a.SchedulerImpl] Process[es] is up
sonarqube-sonarqube-5454cd7496-hfp9c sonarqube 2022.11.29 17:45:57 INFO  app[][o.s.a.ProcessLauncherImpl] Launch process[WEB_SERVER] from [/opt/sonarqube]: /usr/lib/jvm/java-11-openjdk/bin/java -Djava.awt.headless=true -Dfile.encoding=UTF-8 -Djava.io.tmpdir=/opt/sonarqube/temp -XX:-OmitStackTraceInFastThrow --add-opens=java.base/java.util=ALL-UNNAMED --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.io=ALL-UNNAMED --add-opens=java.rmi/sun.rmi.transport=ALL-UNNAMED --add-exports=java.base/jdk.internal.ref=ALL-UNNAMED --add-opens=java.base/java.nio=ALL-UNNAMED --add-opens=java.base/sun.nio.ch=ALL-UNNAMED --add-opens=java.management/sun.management=ALL-UNNAMED --add-opens=jdk.management/com.sun.management.internal=ALL-UNNAMED -Dcom.redhat.fips=false -javaagent:/opt/sonarqube/data/jmx_prometheus_javaagent.jar=8000:/opt/sonarqube/conf/prometheus-config.yaml -Dhttp.nonProxyHosts=localhost|127.*|[::1] -cp ./lib/sonar-application-9.7.1.62043.jar:/opt/sonarqube/lib/jdbc/postgresql/postgresql-42.4.1.jar org.sonar.server.app.WebServer /opt/sonarqube/temp/sq-process13968265659864058523properties
sonarqube-sonarqube-5454cd7496-hfp9c sonarqube Error opening zip file or JAR manifest missing : /opt/sonarqube/data/jmx_prometheus_javaagent.jar
sonarqube-sonarqube-5454cd7496-hfp9c sonarqube Error occurred during initialization of VM
sonarqube-sonarqube-5454cd7496-hfp9c sonarqube agent library failed to init: instrument
sonarqube-sonarqube-5454cd7496-hfp9c sonarqube 2022.11.29 17:45:57 WARN  app[][o.s.a.p.AbstractManagedProcess] Process exited with exit value [Web Server]: 1
sonarqube-sonarqube-5454cd7496-hfp9c sonarqube 2022.11.29 17:45:57 INFO  app[][o.s.a.SchedulerImpl] Process[Web Server] is stopped
sonarqube-sonarqube-5454cd7496-hfp9c sonarqube 2022.11.29 17:45:57 WARN  app[][o.s.a.p.AbstractManagedProcess] Process exited with exit value [ElasticSearch]: 143
sonarqube-sonarqube-5454cd7496-hfp9c sonarqube 2022.11.29 17:45:57 INFO  app[][o.s.a.SchedulerImpl] Process[ElasticSearch] is stopped
sonarqube-sonarqube-5454cd7496-hfp9c sonarqube 2022.11.29 17:45:57 INFO  app[][o.s.a.SchedulerImpl] SonarQube is stopped
```

The Deployment manifest is missing some declarations from the Stefulset manifest, for example :

- Prometheus JMX exporter
- Wait-for-db init container  
 
Almost all changes come from a difference between the Stafulset and Deployment manifest.

It works well from my side, and Prometheus can now fetch the 3 metrics endpoints with a classic PodMonitor definition.

Any clue why the deployment manifest is way behind the Statefulset one? 

Please ensure your pull request adheres to the following guidelines:
- [x] explain your motives to contribute this change: what problem you are trying to fix, what improvement you are trying to make
- [x] Document your Changes in the `CHANGELOG.md` file of the respected chart as well as the `Chart.yaml`
- [x] Bump the Version number of the respected chart

jCOTINEAU commented:
hello @mydoomfr thanks a lot for taking some time to participate in the community.

It looks like your branch is not up to date with our master, could you please sync and rebase it ? 

As so far, I am not able to test your branch , the helm dependencies does not build anymore with ```https://charts.bitnami.com/bitnami``` and postgresql version ```10.0.15```

Also when changing that repo to ```https://raw.githubusercontent.com/bitnami/charts/pre-2022/bitnami``` and templating with default values + running our tests, i get that error message 

```
Error: parse error at (sonarqube/templates/deployment.yaml:493): unexpected {{end}}
helm.go:84: [debug] parse error at (sonarqube/templates/deployment.yaml:493): unexpected {{end}}
```

Do you get the same result by running the test suite described [here](https://github.com/mydoomfr/helm-chart-sonarqube/blob/master/tests/unit-compatibility-test/README.md) ?

To answer your question, right now stateful set is our default way of running SonarQube, which might explain why it got a little bit more attention.

Waiting forwards for your update.

mydoomfr commented:
I just updated my branch and rebased it from master, it should be good now, and I re-tested the Deployment. I got the same error as you faced and fixed it. Should be good now.

```
Error: parse error at (sonarqube/templates/deployment.yaml:493): unexpected {{end}}
helm.go:84: [debug] parse error at (sonarqube/templates/deployment.yaml:493): unexpected {{end}}
```

I omitted the `init-fs` init-container from the "Statefulset" because in "Deployment," it probably makes no sense?

## Tests
<details>

```console
$ charts/sonarqube
$ ../../.cirrus/unit_helm_compatibility_test.sh

Running unit compatibility tests for Kubernetes version 1.25.0
Entering test for change-admin-password-hook-values.yaml
install.go:192: [debug] Original chart version: ""
install.go:209: [debug] CHART PATH: /Users/pinchonbenjamin/Code/github/mydoomfr/helm-chart-sonarqube/charts/sonarqube

Summary: 18 resources found parsing stdin - Valid: 18, Invalid: 0, Errors: 0, Skipped: 0
Ending test for change-admin-password-hook-values.yaml
Entering test for config-values.yaml
install.go:192: [debug] Original chart version: ""
install.go:209: [debug] CHART PATH: /Users/pinchonbenjamin/Code/github/mydoomfr/helm-chart-sonarqube/charts/sonarqube

Summary: 16 resources found parsing stdin - Valid: 16, Invalid: 0, Errors: 0, Skipped: 0
Ending test for config-values.yaml
Entering test for default-values.yaml
install.go:192: [debug] Original chart version: ""
install.go:209: [debug] CHART PATH: /Users/pinchonbenjamin/Code/github/mydoomfr/helm-chart-sonarqube/charts/sonarqube

Summary: 16 resources found parsing stdin - Valid: 16, Invalid: 0, Errors: 0, Skipped: 0
Ending test for default-values.yaml
Entering test for deployment-values.yaml
install.go:192: [debug] Original chart version: ""
install.go:209: [debug] CHART PATH: /Users/pinchonbenjamin/Code/github/mydoomfr/helm-chart-sonarqube/charts/sonarqube

Summary: 16 resources found parsing stdin - Valid: 16, Invalid: 0, Errors: 0, Skipped: 0
Ending test for deployment-values.yaml
Entering test for ingress-values.yaml
install.go:192: [debug] Original chart version: ""
install.go:209: [debug] CHART PATH: /Users/pinchonbenjamin/Code/github/mydoomfr/helm-chart-sonarqube/charts/sonarqube

Summary: 17 resources found parsing stdin - Valid: 17, Invalid: 0, Errors: 0, Skipped: 0
Ending test for ingress-values.yaml
Entering test for install-plugins-values.yaml
install.go:192: [debug] Original chart version: ""
install.go:209: [debug] CHART PATH: /Users/pinchonbenjamin/Code/github/mydoomfr/helm-chart-sonarqube/charts/sonarqube

Summary: 16 resources found parsing stdin - Valid: 16, Invalid: 0, Errors: 0, Skipped: 0
Ending test for install-plugins-values.yaml
Entering test for jdbc-config-values.yaml
install.go:192: [debug] Original chart version: ""
install.go:209: [debug] CHART PATH: /Users/pinchonbenjamin/Code/github/mydoomfr/helm-chart-sonarqube/charts/sonarqube

Summary: 16 resources found parsing stdin - Valid: 16, Invalid: 0, Errors: 0, Skipped: 0
Ending test for jdbc-config-values.yaml
Entering test for networkpolicy-values.yaml
install.go:192: [debug] Original chart version: ""
install.go:209: [debug] CHART PATH: /Users/pinchonbenjamin/Code/github/mydoomfr/helm-chart-sonarqube/charts/sonarqube

Summary: 19 resources found parsing stdin - Valid: 19, Invalid: 0, Errors: 0, Skipped: 0
Ending test for networkpolicy-values.yaml
Entering test for pvc-values.yaml
install.go:192: [debug] Original chart version: ""
install.go:209: [debug] CHART PATH: /Users/pinchonbenjamin/Code/github/mydoomfr/helm-chart-sonarqube/charts/sonarqube

Summary: 17 resources found parsing stdin - Valid: 17, Invalid: 0, Errors: 0, Skipped: 0
Ending test for pvc-values.yaml
Entering test for secret-values.yaml
install.go:192: [debug] Original chart version: ""
install.go:209: [debug] CHART PATH: /Users/pinchonbenjamin/Code/github/mydoomfr/helm-chart-sonarqube/charts/sonarqube

Summary: 15 resources found parsing stdin - Valid: 15, Invalid: 0, Errors: 0, Skipped: 0
Ending test for secret-values.yaml
Entering test for service-values.yaml
install.go:192: [debug] Original chart version: ""
install.go:209: [debug] CHART PATH: /Users/pinchonbenjamin/Code/github/mydoomfr/helm-chart-sonarqube/charts/sonarqube

Summary: 16 resources found parsing stdin - Valid: 16, Invalid: 0, Errors: 0, Skipped: 0
Ending test for service-values.yaml
Entering test for serviceaccount-values.yaml
install.go:192: [debug] Original chart version: ""
install.go:209: [debug] CHART PATH: /Users/pinchonbenjamin/Code/github/mydoomfr/helm-chart-sonarqube/charts/sonarqube

Summary: 17 resources found parsing stdin - Valid: 17, Invalid: 0, Errors: 0, Skipped: 0
Ending test for serviceaccount-values.yaml
```

</details>

## Deploy

```console
$ helm upgrade --install -n sonarqube sonarqube . -f my-values.yaml
```

```console
$ kubectl get po -n sonarqube

NAME                                  READY   STATUS    RESTARTS   AGE
sonarqube-postgresql-0                1/1     Running   0          2m46s
sonarqube-sonarqube-5f75fbfd4-s294v   1/1     Running   0          2m46s
```

## Podmonitor

```yaml
---
apiVersion: monitoring.coreos.com/v1
kind: PodMonitor
metadata:
  name: sonarqube-metrics
spec:
  selector:
    matchLabels:
      app: sonarqube
  namespaceSelector:
    matchNames:
    - sonarqube
  podMetricsEndpoints:
  - interval: 30s
    path: /api/monitoring/metrics
    scheme: http
    port: http
    bearerTokenSecret:
      name: sonarqube-sonarqube-monitoring-passcode
      key: SONAR_WEB_SYSTEMPASSCODE
  - interval: 30s
    path: /
    scheme: http
    port: monitoring-ce
  - interval: 30s
    path: /
    scheme: http
    port: monitoring-web
```

<img width="605" alt="image" src="https://user-images.githubusercontent.com/1282754/210751151-69ba6d37-33e9-46d7-9f64-d88a02c568fa.png">

jCOTINEAU commented:
Hello @mydoomfr i just want to give you an update on that PR.

We have a new chart LTS release coming up.
Which is the reason why we are not yet incorporating your changes.

I deeply thank you for taking the time to participate in the community regarding the time it takes for the PR to be accepted.

I would nonetheless like to highlight from our [CONTRIBUTION.md](https://github.com/SonarSource/helm-chart-sonarqube#contributing) that it is hard for us to deal with external PR and we always prefer feature demand on our community forum.

Huge thanks again, may you excuse us for the inconvenience

jCOTINEAU commented:
Hello @mydoomfr 

thanks again for your contribution. As said, before, we will consider your change after the LTS is released, in one of the next hardening sprints.

I have just written a small task about it ([SQHELM-108](https://sonarsource.atlassian.net/browse/SQHELM-108)) and we'll make sure to tackle it using the work you did on this feature branch.

For now, I'm closing this PR.
